torch/fx/experimental/Partitioner.py

```diff
@@ -106,77 +122,378 @@ def get_output_nodes(self) -> List[Node]:
+    a RuntimeError is raised.
+    If the whole graph cannot fit into one devices but
+    could be split into multiple devices, it goes to self.size_based_partition.
+    After the size_based_partition, it checks if the number of partitions exceeds
+    the number of devices. If it does, a RuntimeError is raised.
+    Otherwise, a DAG structure is returned
+    along with a new graph module with partitions as submodule nodes.
     """
     def __init__(self) -> None:
-        """
-        After a partitioner is created,
-        it first check if multiple types of devices (backends) are involved.
-        So far, we assume that there is only one backend and one device with unlimited memory
-        """
-        self.partitions: List[Partition] = []
-        self.devices: List[Dict[str, Union[str, float]]] = []
+        self.partitions: Set[Partition] = set()
+        self.devices: List[Device] = []
         self.node_to_partitions: Dict[Node, List[int]] = {}
+        self.partition_to_used_mem: Dict[Partition, int] = {}
+        self.orphans: Set[Node] = set()
+        self.args_mapping: Dict[Node, List[Node]] = {}
+        self.users_mapping: Dict[Node, List[Node]] = {}
 
-    def partition_graph(self, fx_module: GraphModule, devices: List[Dict[str, Union[str, float]]]) -> DAG:
+    def partition_graph(
+        self,
+        fx_module: GraphModule,
+        torch_module: torch.nn.Module,
+        devices: List[Device]
+    ) -> PartitionResult:
         """
-        Given the fx_module and devices, find the partitions, do the partitions,
-        and then return a dictionary representing the DAG of partitions
+        Given the fx module, torch module and devices,
+        find the partitions, do the partitions,
+        and then return a DAG and a new fx module with submodule nodes (partitions)
         """
         self.graph_module = fx_module
         self.devices = devices
-        # Create a dummy partition as root.
-        self.root_partition = self.create_partition()
-        # So far, the whole could fit into one device since we assume the memory is unlimited.
-        # TODO: Check the available memory size for each device and see if one device is enough or multiple partitions are needed.
-        self.find_single_partition()
-        self.do_partition()
-        dag = self.dump_partition_DAG()
-        return dag
+        self.torch_module = torch_module
+        if len(self.devices) == 0:
+            raise RuntimeError('No devices')
+        available_mem = self.devices[0].available_mem_bytes
+        # Check if there are op nodes in the graph
+        has_op_nodes = False
+        for node in self.graph_module.graph.nodes:
+            if node.op == 'output':
+                break
+            if node.op not in {'placeholder', 'get_attr'}:
+                has_op_nodes = True
+                break
+        if not has_op_nodes:
+            raise RuntimeError('No Partition since no operations in the module')
```

Comment:
I think you could write this whole check more concisely as
```
nodes = self.graph_module.graph.nodes
if not any(node.op not in {'placeholder', 'get_attr', 'output'} for node in nodes):
            raise RuntimeError('No Partition since no operations in the module')
```

Alternatively you can further simplify it via DeMorgan's law

```
nodes = self.graph_module.graph.nodes
if all(node.op in {'placeholder', 'get_attr', 'output'} for node in nodes):
            raise RuntimeError('No Partition since no operations in the module')
```

You don't need to worry about breaking on `output` node, you can assume the Graph is well-formed at the input